Change auto completion scopes

NobodysNightmare opened the pull request:
## OpenProject work packages

https://community.openproject.org/work_packages/17549
## Description

This PR supports "project free" autocompletion as required in https://github.com/opf/openproject/pull/3124.

It seems like some history is involved here:
The auto completion endpoint was initially developed for the use case of autocompleting related work packages. Therefore the endpoint would only return work packages that can go into relation with work packages for a given project (even with a scope called "all"). However, in the meantime new use cases for autocompletion arose (autocompletion in text fields, e.g. description), for those cases the limitation to relatable fields did not make much sense.

This PR makes the following changes to the endpoint `work_packages/auto_complete.json`:
- the `project_id` is now optional
- the scope called `all` was renamed to `relatable`
- a new scope called `all` was introduced, meaning "from everywhere"
### Behavioural comparison

| scope | behaviour before | behaviour now |
| --- | --- | --- |
| no scope | always return WPs from given project | return WPs from given project if a project is given, return WPs from _everywhere_ otherwise |
| `all` | if cross-project relations are allowed return from WPs from _everywhere_, from the given project otherwise | always return WPs from _everywhere_ |
| `relatable` | always return WPs from given project (option was not implemented) | like the **old** behaviour of `all` |
### Note

This PR is intended for **Release 4.3**, currently there is no distinct branch for that release, so this PR is targetting the `dev` branch.

NobodysNightmare committed:
rename query to scope

sounds more understandable to me, but I am biased, I guess...

NobodysNightmare committed:
use more clear scopes for autocompletion

lets see if that really makes stuff more understandable...
I (personally) found the old definition of "all" to be too narrow.

Now there is:

scope == "all": I want to search over all WPs
scope == "relatable": I want to search over all WPs that are relatable to WPs in the given project (via project_id)
no scope: If I specified a project search over the WPs of that one, otherwise search everywhere (like "all")

NobodysNightmare committed:
explain why some code exists

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-autocomplete-scopes.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-autocomplete-scopes.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

0xF013 commented:
no scope lol

Azure7111 commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -37,11 +37,23 @@ def index
 
-      @work_packages |= query.visible.find_all_by_id(q.to_i) if q =~ /\A\d+\z/
+        scope = Setting.cross_project_work_package_relations? ? WorkPackage : @project.work_packages
+      elsif @project.nil? || params[:scope] == 'all'
```
Won't the line above break if ``@project == nil``?

Azure7111 commented on spec/controllers/work_packages/auto_completes_controller_spec.rb:
```diff
@@ -203,38 +203,74 @@
-        it_behaves_like 'successful response'
+        context 'with scope "all"' do
+          let(:scope) { 'all' }
+          let(:expected_values) { work_package_4 }
```
Although this test is perfectly fine technically, for this case I'd add another work_package (which is available at least in the relatable test as well) since `relatable` is a subset of `all` and hence there is a work_package which has to be left out. Just to be extra safe.

NobodysNightmare replied:
I would agree if there was a check that non-expected values are actually not part of the response.

If I was serious I should probably rewrite all the specs here... But I am really not in the mood, TBH.

Azure7111 replied:
I'd have thought that with my suggestion you could implicitly do that exact check without much hassle.

Though if you really have to rewrite all the specs to make this possible, then this is probably something for another point in time.

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -37,11 +37,23 @@ def index
+      # query for exact ID matches first, to make an exact match the first result of autocompletion
+      @work_packages |= scope.visible.find_all_by_id(q.to_i) if q =~ /\A\d+\z/
+
+      @work_packages |= scope.visible.find(:all,
```
This is very hard to read (3 layers of if statements). Any chance to refactor @NobodysNightmare?

myabc replied:
@NobodysNightmare not sure when this (or #2104 will get merged in), but might also be a good opportunity to get rid of the `find(:all, ..` with conditions hash (deprecated in Rails 4.0, removed in Rails 4.1).

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -37,11 +37,23 @@ def index
-      @work_packages |= query.visible.find_all_by_id(q.to_i) if q =~ /\A\d+\z/
+        scope = Setting.cross_project_work_package_relations? ? WorkPackage : @project.work_packages
+      elsif @project.nil? || params[:scope] == 'all'
+        scope = WorkPackage
```
Didn't you want to use `WorkPackage.scoped` here?

NobodysNightmare committed:
try to increase readability

- extract scopedeterminition into an extra method
- always use "outer" if statements in index method

NobodysNightmare committed:
always return a valid scope

though WorkPackage works like a scope *most* of the time, it is not a scope

e.g. you can't directly call :order on WorkPackage

Note: I would have loved to already do it the Rails 4 way and call :all
However in Rails 3 :all returns an Array and not a Relation, so we will
have to wait for the Rails 4(.1) Upgrade to replace :scoped with :all :(

Sorry @myabc

NobodysNightmare committed:
rewrite find into where, order and limit

NobodysNightmare committed:
increase readability of wp->hash conversion

it might use more lines, but I hope there is a gain in readability

NobodysNightmare commented:
I think I adressed the issues mentioned, I'd aprechiate a re-review...

NobodysNightmare commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -34,39 +34,63 @@ class WorkPackages::AutoCompletesController < ApplicationController
+      @work_packages |= scope.visible
+                             .where(sql_query)
+                             .order(id: :asc)
+                             .limit(10)
```
@houndci All arguments are aligned beautifully here, what's your point?

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -34,39 +34,63 @@ class WorkPackages::AutoCompletesController < ApplicationController
+    else
+      @project.work_packages
+    end
   end
```
If I have looked at it correctly, your changes now means that `@project` is only referenced inside `determine_scope` which is good. 

But you no longer need an instance variable, then.

Following this line of thought, you no longer need the before filter, then.

> You can get it if you really want
> You can get it if you really want
> You can get it if you really want
> But you must try, try and try
> Try and try, you'll succeed at last

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -34,39 +34,63 @@ class WorkPackages::AutoCompletesController < ApplicationController
+    @project = Project.find(project_id) if project_id
   rescue ActiveRecord::RecordNotFound
-    render_404
+    @project = nil
```
`Project.find_by_id` will give you the same result without having to rescue from not finding the record yourself.

NobodysNightmare replied:
Actually passing the `identifier` instead of the `id` is currently also valid and something that happens.

NobodysNightmare committed:
eliminate instance variable

NobodysNightmare committed:
also accept auto complete requests without query

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -30,43 +30,65 @@
-    end
+    @work_packages |= scope.visible
+                           .where(sql_query)
+                           .order(id: :asc)
```
<pre>

ActiveRecord::StatementInvalid (PG::SyntaxError: ERROR:  non-integer constant in ORDER BY
LINE 2: ...AST(work_packages.id AS CHAR(13)) LIKE '%45%') ORDER BY '---
                                                                   ^
: SELECT  "work_packages"."id" AS t0_r0, "work_packages"."type_id" AS t0_r1, "work_packages"."project_id" AS t0_r2, "work_packages"."subject" AS t0_r3, "work_packages"."description" AS t0_r4, "work_packages"."due_date" AS t0_r5, "work_packages"."category_id" AS t0_r6, "work_packages"."status_id" AS t0_r7, "work_packages"."assigned_to_id" AS t0_r8, "work_packages"."priority_id" AS t0_r9, "work_packages"."fixed_version_id" AS t0_r10, "work_packages"."author_id" AS t0_r11, "work_packages"."lock_version" AS t0_r12, "work_packages"."done_ratio" AS t0_r13, "work_packages"."estimated_hours" AS t0_r14, "work_packages"."created_at" AS t0_r15, "work_packages"."updated_at" AS t0_r16, "work_packages"."start_date" AS t0_r17, "work_packages"."parent_id" AS t0_r18, "work_packages"."responsible_id" AS t0_r19, "work_packages"."root_id" AS t0_r20, "work_packages"."lft" AS t0_r21, "work_packages"."rgt" AS t0_r22, "projects"."id" AS t1_r0, "projects"."name" AS t1_r1, "projects"."description" AS t1_r2, "projects"."homepage" AS t1_r3, "projects"."is_public" AS t1_r4, "projects"."parent_id" AS t1_r5, "projects"."created_on" AS t1_r6, "projects"."updated_on" AS t1_r7, "projects"."identifier" AS t1_r8, "projects"."status" AS t1_r9, "projects"."lft" AS t1_r10, "projects"."rgt" AS t1_r11, "projects"."summary" AS t1_r12, "projects"."project_type_id" AS t1_r13, "projects"."responsible_id" AS t1_r14, "projects"."work_packages_responsible_id" AS t1_r15 FROM "work_packages" LEFT OUTER JOIN "projects" ON "projects"."id" = "work_packages"."project_id" WHERE (projects.status=1 AND projects.id IN (SELECT em.project_id FROM enabled_modules em WHERE em.name='work_package_tracking')) AND (LOWER(work_packages.subject) LIKE '%45%' OR
                    CAST(work_packages.id AS CHAR(13)) LIKE '%45%') ORDER BY '---
:id: :asc
' LIMIT 10):
  app/controllers/work_packages/auto_completes_controller.rb:58:in `|'
  app/controllers/work_packages/auto_completes_controller.rb:58:in `index'
  app/middleware/params_parser_with_exclusion.rb:40:in `call'

</pre>

ulferts replied:
http://apidock.com/rails/ActiveRecord/QueryMethods/order#1542-Order-with-hash-parameters-only-in-ActiveRecord-4-

NobodysNightmare committed:
specify order in the Rails 3 way

ascending is the default anyway

ulferts commented on app/controllers/work_packages/auto_completes_controller.rb:
```diff
@@ -30,43 +30,65 @@
-    end
+    @work_packages |= scope.visible
+                           .where(sql_query)
+                           .order(:id)
```
lol

<pre>
ActiveRecord::StatementInvalid (PG::AmbiguousColumn: ERROR:  column reference "id" is ambiguous
LINE 2: ...rk_packages.id AS CHAR(13)) LIKE '%65%') ORDER BY id LIMIT 1...
                                                             ^
: SELECT  "work_packages"."id" AS t0_r0, "work_packages"."type_id" AS t0_r1, "work_packages"."project_id" AS t0_r2, "work_packages"."subject" AS t0_r3, "work_packages"."description" AS t0_r4, "work_packages"."due_date" AS t0_r5, "work_packages"."category_id" AS t0_r6, "work_packages"."status_id" AS t0_r7, "work_packages"."assigned_to_id" AS t0_r8, "work_packages"."priority_id" AS t0_r9, "work_packages"."fixed_version_id" AS t0_r10, "work_packages"."author_id" AS t0_r11, "work_packages"."lock_version" AS t0_r12, "work_packages"."done_ratio" AS t0_r13, "work_packages"."estimated_hours" AS t0_r14, "work_packages"."created_at" AS t0_r15, "work_packages"."updated_at" AS t0_r16, "work_packages"."start_date" AS t0_r17, "work_packages"."parent_id" AS t0_r18, "work_packages"."responsible_id" AS t0_r19, "work_packages"."root_id" AS t0_r20, "work_packages"."lft" AS t0_r21, "work_packages"."rgt" AS t0_r22, "projects"."id" AS t1_r0, "projects"."name" AS t1_r1, "projects"."description" AS t1_r2, "projects"."homepage" AS t1_r3, "projects"."is_public" AS t1_r4, "projects"."parent_id" AS t1_r5, "projects"."created_on" AS t1_r6, "projects"."updated_on" AS t1_r7, "projects"."identifier" AS t1_r8, "projects"."status" AS t1_r9, "projects"."lft" AS t1_r10, "projects"."rgt" AS t1_r11, "projects"."summary" AS t1_r12, "projects"."project_type_id" AS t1_r13, "projects"."responsible_id" AS t1_r14, "projects"."work_packages_responsible_id" AS t1_r15 FROM "work_packages" LEFT OUTER JOIN "projects" ON "projects"."id" = "work_packages"."project_id" WHERE (projects.status=1 AND projects.id IN (SELECT em.project_id FROM enabled_modules em WHERE em.name='work_package_tracking')) AND (LOWER(work_packages.subject) LIKE '%65%' OR
                  CAST(work_packages.id AS CHAR(13)) LIKE '%65%') ORDER BY id LIMIT 10):
</pre>

NobodysNightmare committed:
uglify order statement

at least this one DOES work...

ulferts commented:
@NobodysNightmare looks good now...

If it weren't for the relations tab on work packages where the `relatable` scope should also be used.

NobodysNightmare committed:
use "relatable" scope for relations

NobodysNightmare committed:
centralize knowledge of WP autocomplete url

- project information is not needed anymore
- the endpoint format was fixed anyway (also accepted parameters)

NobodysNightmare commented:
@ulferts Lets see what travis has to say.

I fixed the relations autocomplete URL and also took the opportunity to clean up all existing WP autocompletions that source from "all" work packages.

ulferts commented:
@NobodysNightmare looks good so far, please answer my remark but apart from that, it is good to merge. 

The spec failures are not related to this change here.
